### ISSUE TITLE
question about export support

### ISSUE THREAD

**Issue by daquexian** [onnx/onnx-mlir#298]
Thanks for your amazing work!

Is there any plan to support exporting, i.e. translating the mlir dialect back to onnx?

**Comment by AlexandreEichenberger** [comment]
None that is planed. Since MLIR is based on lowering dialects from ONNX to KRNL, and then to affine, SCF, Standard... we are going away from ONNX quite quickly. That is our current focus, namely lower ONNX to binary that can be executed on various platforms.

But maybe you want to implement an ONNX to ONNX optimizer? Do you have any immediate needs in this area? 

There might be functionality in the ONNX converters that easily let you generate an ONNX protobuf file, which could be called from an ONNX MLIR dialect pass. Would you be interested in exploring this?

**Comment by daquexian** [comment]
> But maybe you want to implement an ONNX to ONNX optimizer? Do you have any immediate needs in this area?

Yes, I want to leverage MLIR (especially the DRR) to implement ONNX to ONNX optimizer. It is not urgent. Maybe I can add export support to ONNX-MLIR if I find it is useful to the optimizer.

> There might be functionality in the ONNX converters that easily let you generate an ONNX protobuf file, which could be called from an ONNX MLIR dialect pass.

Could you elaborate? Do you mean translate onnx mlir dialect to onnx protobuf by modifying (change the source format from tensorflow/pytorch etc to onnx dialect) some existing onnx converter? :)

**Comment by tjingrant** [comment]
@daquexian do you intend to rebase https://github.com/daquexian/onnx-simplifier on mlir based infra? That'd be very cool and I'd be glad to help out any way I can!

> Do you mean that translating onnx mlir dialect to onnx protobuf can be achieved by modifying some existing onnx converter? :)

I think Alex meant that you can create a compiler pass inside onnx-mlir operating on ONNX Dialect, and the only thing it does is going through the list of operations in the IR module, translate these operations, one by one, into protobuf object, and thereby getting an isomorphic onnx protobuf representation of the whole model.

I think most of this conversion process (ONNX Dialect -> ONNX Protobuf) can be automated because we automated the other direction of conversion with a reasonable amount of effort. This is something that will for sure benefit a wide audience of the ONNX community. Building onnx-simplifier using the ONNX dialect infrastructure provided here is both natural & synergetic. I'm also available to chat about it online if you are interested, just shoot me an email.
